### PR TITLE
internal/ethapi: add personal_sign method

### PR DESCRIPTION
**Pull request by bas-vk**
This PR includes several things:

1. the behavior of `eth_sign` is changed. It now accepts an arbitrary message, hashes it and calculates the signature of the hash (breaks backwards compatability!).
2. add `personal_sign`, same semantics as `eth_sign` but also accepts a password. The private key used to sign the hash is temporary unlocked in the scope of the request.
3. add `personal_recover`, returns the address for the account that created the signature.

`personal_recover(message, signature)`
`personal_sign(hash, address [, password])`

@alexvandesande

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Aug 24 09:34:35 UTC 2016_

**Comment by Arachnid** [comment]
With the impending advent of hardware wallets, I'm concerned this is overly internal-account-specific; is there any way to avoid hardcoding in passwords as an authentication mechanism? (I don't have a suggestion just yet, just thinking out loud)

**Comment by fjl** [comment]
If you use a hardware wallet, it's better to talk to it on the client side.

**Comment by fjl** [comment]
@bas-vk Can we seize this opportunity to fix some of the other issues with eth_sign?

* It should hash the input on the server side.
* There was some issue with the Ethereum-specific +27 v value

**Comment by Arachnid** [comment]
> If you use a hardware wallet, it's better to talk to it on the client side.

For signing, perhaps, but I'm thinking more about what sort of interfaces we offer that are backed by accounts, and how those will work if we support accounts other than those whose private keys are held by Geth.

**Comment by bas-vk** [comment]
> @bas-vk Can we seize this opportunity to fix some of the other issues with eth_sign?

We can, this PR is rather small so I have no problem adding additional changes.

> - It should hash the input on the server side.

Changing this would break backwards compatibility for `eth_sign`. Exploiting this requires the key to be unlocked. And in that case obtaining the private key is nice but if it's possible to sign arbitrary data a thief  basically has full control over the account during the time he can sign data. I will make this change for the new method. 

> - There was some issue with the Ethereum-specific +27 v value

I believe the issue was that according to the yellow paper Ethereum hashes should have a signature with v+27 just like bitcoin. `eth_sign` currently doesn't add 27. Will add 27 to the signature of the new method to make it compliant with the yellow paper and leave `eth_sign` untouched.

Maybe at some point we can mark `eth_sign` deprecated.

**Comment by fjl** [comment]
Thanks. I agree that eth_sign should be deprecated, ideally quite soon. We can keep it around for the 1.5  release cycle and remove it in 1.6.

**Comment by kumavis** [comment]
@fjl does the api take the hash or the message?
`personal_sign(messageHash, signerAddress [, password])`
`personal_sign(messageBody, signerAddress [, password])`

Its important that we pass the message so that the id mgmt layer (mist/metamask) can present the message to the user for approval

**Comment by fjl** [comment]
My proposal is as follows:

`personal_sign(message, address [, password])` should sign keccak256(message) and sign it with the key that belongs to the address.

In internal discussions we have considered to prepend "Signed Ethereum message:\n" to the message. This would make it much harder to misuse the API. The idea of adding the prefix is not set in stone though.

**Comment by kumavis** [comment]
I'm not sure how the prefix makes it much harder to misuse the API. can you expand?
If the attacker is trying to find a hash collision with a tx sig, they would just build a rainbow table with that prefix as the 'salt'.

I understand how a salt might be better than none, but maybe a dynamic salt like a timestamp would provide more security.

**Comment by fjl** [comment]
What we are concerned with is the case of a malicious dapp creating (transaction) signatures on behalf of inexperienced users who just click yes.

**Comment by kumavis** [comment]
Yes, that is the primary attack vector im worried about.

**Comment by fjl** [comment]
If you have more ideas for making this safer please dump them here. Adding a timestamp sounds nice at first, but keep in mind that something (probably a contract) will need to verify the signature eventually and it needs to add the same prefix. The user would need to submit the timestamp to said contract later and that sounds complicated. But maybe it's the right thing to do.

**Comment by kumavis** [comment]
FWIW we are currently using the old version of the `eth_sign` API and displaying the messages.
We're currently reworking this UI and it will be more clear what exactly you are signing, the size of the message, and have a bunch of warnings not to sign any weird messages.

image

**Comment by kumavis** [comment]
or just use a random salt and include it in the serialized message signature
```
personal_sign( message ){
  salt = random()
  sig = sign( hash( salt + message ) )
  return [sig, salt]
}
```

**Comment by fjl** [comment]
Yes that looks better. I'll think about it.

**Comment by kumavis** [comment]
requesting comment from @Georgi87 and @mhhf who make use of user message signatures in the https://www.gnosis.pm/ predication market dapp

**Comment by kumavis** [comment]
Theres still a problem with signing messages with the same key we use for tx sigs.

given a **message signature**, theres still a large search space for valid tx hashes that could match the message signature. Tx parameters such as `value` and `toAddress` provide a lot of search space that can be explored in parallel to find a hash collision.
```
{ evilTx, value: 0 wei }
{ evilTx, value: 1wei }
{ evilTx, value: 2wei }
...
```

**Comment by fjl** [comment]
@nagydani would be good to have your input as well

**Comment by flyswatter** [comment]
Since we're talking about `eth_sign`, I'd like to chime in here that I think it would be way more useful if a method allowed signing/decrypting arbitrary lengthed data, not just hashes, and [I've expressed this opinion in an EIP that hasn't gotten much of a response](https://github.com/ethereum/EIPs/issues/130).

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 85bfe8382d06928d2bb2aa14d7957ab0e9dc879f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 970251c18c44ebfc8faae67242f8c82184fcc823
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 85bfe8382d06928d2bb2aa14d7957ab0e9dc879f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 42a26000eacd23ca325623c6460a70f5c8d4a50f
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by nagydani** [comment]
I have a problem with `SignEthereum`. It can still be used to extract a private key by supplying it with a "hash" that is not actually a hash of something. Wouldn't it be better to make the message its argument and let it do the hashing internally?

**Comment by bas-vk** [comment]
`crypto.SignEthereum` is a low level function and not accessible through the RPC. Currently it is up to the caller to hash the message as happens in the RPC/API.  We could change this but it limits its use to only be able to sign hashed message. The same thing applies to the existing `crypto.Sign` which is for example used in p2p to sign something else.

**Comment by nagydani** [comment]
Would it make sense to annotate in comments the functions that are exposed through RPC/API?

**Comment by Arachnid** [comment]
> or just use a random salt and include it in the serialized message signature

The problem with this is that you can't be sure that there isn't some message that is meaningful to some system when signed and permits an arbitrary value first field. Prepending an unambiguous message prevents these sort of attacks.

> given a message signature, theres still a large search space for valid tx hashes that could match the message signature.

Are you talking about finding collisions in SHA3? I don't think this is worth worrying about.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 85bfe8382d06928d2bb2aa14d7957ab0e9dc879f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 42a26000eacd23ca325623c6460a70f5c8d4a50f
  * Commits must be prefixed with the package(s) they modify

 * Commit: f116d1ab3553dd75ddbef69658772d6cd6407151
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by bas-vk** [comment]
> Would it make sense to annotate in comments the functions that are exposed through RPC/API?

I don't think its very useful to add this to the documentation of these function. I did add a warning to the docs that using these 2 sign functions securely is the responsibility of the caller.

> The problem with this is that you can't be sure that there isn't some message that is meaningful to some system when signed and permits an arbitrary value first field. Prepending an unambiguous message prevents these sort of attacks.

We can use something similar as Bitcoin.
`\x18Bitcoin Signed Message:\n#{message.size.chr}#{message}`

**Comment by pipermerriam** [comment]
Would like to voice my opposition to restricting this API such that it cannot be used to sign transactions.  One valid use case for using the `personal_sign` for txn signing.

* I wish to send a transaction but I do not have the chain fully synced.
* I can spin up the RPC server and use `personal_sign` to create a signed transaction.
* I can then submit that transaction over a public RPC server through the `eth_sendRawTransaction` API.

Thoughts?

**Comment by kumavis** [comment]
> Prepending an unambiguous message prevents these sort of attacks.

hmm. I can't quite imagine a practical example of the attack, but get the idea.

> Are you talking about finding collisions in SHA3? I don't think this is worth worrying about.

Well then it would seem the only attack vector is the user signing a non-human intelligible unicode-encoded tx. I'm less worried about that bc we will show lots of warnings not to sign weird messages. We could also do a test to see if the message deserialized as a tx.

> I can spin up the RPC server and use personal_sign to create a signed transaction.

If your rpc provider can perform `personal_sign` and `eth_sendRawTransaction` why not just sign with `eth_sendTransaction` ?

**Comment by Georgi87** [comment]
Checking this request from Gnosis's perspective. We are using eth_sign quite a lot for communications with our API and signing oracle results but switching to `personal_sign` would be not a problem.

I also favor `personal_sign` and support that `eth_sign` should be deprecated.

> arbitrary message
I guess we would still have to limit it maybe to 1MB? We should not make the assumption that geth or other nodes handle very large files for singing but should advise sending only hashes. So hashes of hashes would be signed.

Not sure yet about the random salt. It prevents to sign any serialized transaction but maybe we should keep this as an option? It should be the wallet's responsibility to warn users about potentially signed transactions.

I have bad feelings sending the cleartext password as part of an RPC request. Why not assuming an unlocked account as it is currently with `eth_sign`?

**Comment by fjl** [comment]
> I have mixed feelings sending (optionally) the cleartext password as part of an RPC request. Why not assuming an unlocked account as it is currently with eth_sign

Account unlocking is an insecure concept. We are trying to move away from that concept as much as possible. In the ideal case geth would not handle account keys at all. Unfortunately this is hard to do because everyone relies on the APIs as they are now.

Dapps don't need to care about the password, it will be handled by Mist/MetaMask.

**Comment by flyswatter** [comment]
If a new sign method allows larger data blobs, I would say let the RPC specify its own size limit, if an advanced dapp later on requires signing something huge, people can run their own wallet/RPC pairs.

**Comment by flyswatter** [comment]
I'd also like to express my general agreement with @pipermerriam's point, that unrestricted signing is very powerful, and should be exposed in some form.

That said, maybe a method like a `safe_sign` method could also exist, that includes a salt prefix, to help avoid tx phishing attacks.

This would make it easy for wallets like Mist or MetaMask to represent to the user the level of risk associated with a given signature.

**Comment by bas-vk** [comment]
Okay, first of all its important to understand that exploiting the existing `eth_sign` needs an unlocked account and a running RPC interface that is accessible by an adversary. If these 2 conditions are met `eth_sendTransaction` can already be used to send transaction and thus steal funds or send transaction to contracts. We have been telling people to not run a public RPC interface in combination with unlocked accounts since it imposes a security risk.

Proposal:
`eth_sign`: always prefix the given message with `\x18Ethereum Signed Message:\n#{message.size.chr}#{message}` and hash before signing. This prevents the discussed attacks.

`personal_sign`: is not exposed by default and requires the password as argument. Abusing this method will require the victim to run the RPC on a public interface, enable the `personal` namespace and the attacker needs to know the password for an account to abuse. Therefore I think it's okay to give the user more options to relax the security constraints and make it more generic usable. Let this method accept an "options" argument. This argument has the following fields:
```
{
   prefix: string  // optional, prefix message before signing
   hash: bool      // optional, hash message before signing
}
```

Defaults:
```
prefix: \x18Ethereum Signed Message:\n#{message.size.chr}#{message}
hash: true
```
This means the default behavior is just like `eth_sign`.
`sign(keccak256($prefix || $message))`

Users can allow any "nonce" or prefix, even "" which will allow transaction signing. We could omit the hash field and hash always to prevent the attack @nagydani mentioned. But if the personal namespace is already accessible an attacker can unlock account with `personal_unlockAccount` and do `eth_sendTransaction`. Signing arbitrary message is more general than sending transaction but I think most attacks will be about stealing funds are sending transactions to contracts.

**Comment by fjl** [comment]
The hash cannot be optional. The prefix should not be optional, but I'd give in if there is overwhelming demand for a prefix-less version. That would shift more responsibility to the approval UI.

**Comment by bas-vk** [comment]
Would be nice if we get some additional feedback on:
1. should we make the hashing optional in `personal_sign`?
2. should we make the prefix optional in `personal_sign`?

**Comment by Arachnid** [comment]
> Account unlocking is an insecure concept. We are trying to move away from that concept as much as possible.

+1000 on that - but I don't much like an intermediate step that hardcodes passwords as an authentication mechanism - and encourages users to type them into third-party apps calling geth APIs, too!

> 1. should we make the hashing optional in personal_sign?

It seems clear that this exposes a significant security risk of key recovery, and definitely shouldn't be allowed.

> 2. should we make the prefix optional in personal_sign?

I think this also shouldn't be allowed, since it permits other attacks against the user. I haven't heard any concrete use-cases that this would prohibit, either.

**Comment by zelig** [comment]
I think both hashing and prefixing should be encouranged.
The strongest way is to simply make them mandatory.
Surely this is breaking backward compatibility but since all the production-system stakeholders are here probably :), let's do it and call it a day.

With so much interest here and Martin starting, I really hope we will soon come up with a roadmap 
on account key management and tx signing. This unlock thing and the keystore are a liability.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 85bfe8382d06928d2bb2aa14d7957ab0e9dc879f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 42a26000eacd23ca325623c6460a70f5c8d4a50f
  * Commits must be prefixed with the package(s) they modify

 * Commit: f116d1ab3553dd75ddbef69658772d6cd6407151
  * Commits must be prefixed with the package(s) they modify

 * Commit: 22b2020874eac575a3eab31b9de579a22392ebb4
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 85bfe8382d06928d2bb2aa14d7957ab0e9dc879f
  * Commits must be prefixed with the package(s) they modify

 * Commit: 42a26000eacd23ca325623c6460a70f5c8d4a50f
  * Commits must be prefixed with the package(s) they modify

 * Commit: f116d1ab3553dd75ddbef69658772d6cd6407151
  * Commits must be prefixed with the package(s) they modify

 * Commit: 9ae6a21e4966511a6c6ccb7126a14120b79a6a89
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by bas-vk** [comment]
Okay, the latest commit will always prefix and hash the message before calculating the signature.
PTAL.

**Comment by kumavis** [comment]
looks good to me, minus the gitcop/travis warnings

**Comment by flyswatter** [comment]
I'm much happier with this PR now that it has a separate, non-prefixing `personal_sign` method, and that the `sign` method will automatically hash input instead of requiring hashed input.

I do still think there's space for true encryption exposed on the `web3` object in the future, but I don't think it needs to be settled in this PR.

Good work!

**Comment by fjl** [comment]
Note that `personal_sign` is just as prefixing as `eth_sign` in the final version.

**Comment by kenshyx** [comment]
Hello, any ideas how to `ecrecover` after this update to `eth_sign` because now it doesnt work using this flow https://ethereum.stackexchange.com/questions/1777/workflow-on-signing-a-string-with-private-key-followed-by-signature-verificatio/1794#1794?

**Comment by kumavis** [comment]
@fjl  what is the difference between `personal_sign` and `eth_sign`?

**Comment by pipermerriam** [comment]
> Note that personal_sign is just as prefixing as eth_sign in the final version. We can roll this back, but I'd like to see what the wider community response is.

This is problematic at least for my use case.  The Devcon2 token upgrade has a mechanism for upgrading using a signature from the owner.  This change makes it impossible to produce a valid signature using geth since the smart contract isn't expecting the message to be prefixed.

**Comment by kumavis** [comment]
@pipermerriam prolly easiest to create a new contract that shadows the state of the current one

**Comment by christianlundkvist** [comment]
@Arachnid 
> It seems clear that this exposes a significant security risk of key recovery, and definitely shouldn't be allowed.

Do you know any reference materials/articles about exfiltrating keys by signing chosen raw data? I tried to see if I could compute the private key from the ECDSA algo for the lulz but it wasn't obvious to me.

I also agree with @fjl on this point:

> In the ideal case geth would not handle account keys at all. Unfortunately this is hard to do because everyone relies on the APIs as they are now.

We should be striving to have all key management happen on the client side (in secure hardware whenever possible). For instance in uPort and lightwallet-based projects the ethereum node is only used to fetch data and relay already signed transactions. If we see compelling reasons to change (or remove) APIs we shouldn't be scared to change them IMO. It's still very early days for Ethereum.

**Comment by juanfranblanco** [comment]
This might help you to understand the structure of the prefix if implementing it in another language:

```
   public byte[] HashPrefixedMessage(byte[] message)
        {
            var byteList = new List<byte>();
            var bytePrefix = "0x19".HexToByteArray();
            var textBytePrefix = Encoding.UTF8.GetBytes("Ethereum Signed Message:\n" + message.Length);
           
            byteList.AddRange(bytePrefix);
            byteList.AddRange(textBytePrefix);
            byteList.AddRange(message);
            return Hash(byteList.ToArray());
        }

```

https://github.com/Nethereum/Nethereum/blob/master/src/Nethereum.Web3/Core/EthereumMessageSigner.cs#L20-L30

**Comment by parkan** [comment]
Chiming in from Mediachain:

We've been planning to use `web3.eth.sign` to sign metadata objects to be written off-chain by mediachain nodes (by emitting events). The documentation here:

https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethsign

and here

https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sign

currently implies that 2nd argument is a 32-byte hash. This is perfect for our use case, because `keccak256` is generally not used outside of ethereum (side note: what was the motivation for choosing `keccak256` over `sha256`? was it due to absence of native bitshifts in the EVM?), so we can use sha256 in userspace and pass the 32-byte output to RPC to be signed and the signer should not care. This would allow the signatures to be natively and transparently verified with libp2p-crypto libs (we're adding `secp256k1` support to those)

However, it appears that the [signer](https://github.com/ethereum/go-ethereum/blob/master/internal/ethapi/api.go#L299) (I'm assuming this is the piece that backs `eth_sign`? comments imply this is `personal_sign` but I can't find another signer) actually has a totally different interface, accepts the full data, calculates its own hash using keccak, and prepends a string to the preimage, making any kind of transparent verification of these signatures very awkward.

Am I reading this right? Is the documentation wrong as of 1.5.7? Can I make a plea for retaining the simpler functionality, or at least having a `personal_sign` without a prefix?

**Comment by fjl** [comment]
@parkan We put the restrictions in place to prevent misuse (e.g. unintented signing of transactions) and defend against attacks on the signing function. The restrictions are similar to bitcoind's signmessage API. The documentation needs to be updated.

If you can create signatures without relying on the web3 JSON-RPC API, the restrictions don't apply. Please consider whether there is a way for you to sign locally.

> making any kind of transparent verification of these signatures very awkward.

Verification is simple, just add the same prefix.

**Comment by Georgi87** [comment]
I agree with @pipermerriam and @flyswatter that unrestricted signing should be exposed in some form.

I am working on a Multisig-wallet (https://github.com/ConsenSys/MultiSigWallet/tree/dapp), which allows offline signing of transactions and requires signing without prefix.

From a security point of view, I don't see much of a difference between the current `eth_sendTransaction` and `eth_sign`. In both cases, it is not obvious for the user, what kind of transaction/data he is signing. It is up to the node/user-interface to alert the user about potentially risky actions.

I would favour `personal_sign(hash, address [, password])` with the optional password.

It would be beneficial if there would be a common standard for this kind of request.

**Comment by Arachnid** [comment]
The security issue is that if you can pass in a hash, you can ask the user to sign the null hash, which returns the data required to recover the user's private key. As a result, it's dangerous to expose an API that permits signing arbitrary hashes.

**Comment by pipermerriam** [comment]
So can we instead just permit signing arbitrary data that will be hashed on
the back-end.

**Comment by Georgi87** [comment]
Agree with @pipermerriam 

An alternative could be to exclude the null hash from the range of valid inputs?

**Comment by flyswatter** [comment]
@Arachnid is the null hash the only danger, or would other low or easy to guess values also be dangerous?

**Comment by Arachnid** [comment]
I'm not a cryptographer, but I believe the problem persists in a more general fashion if you can persuade someone to sign arbitrary hashes you provide.

**Comment by Arachnid** [comment]
> So can we instead just permit signing arbitrary data that will be hashed on
the back-end.

The risk of that is the possibility of tricking people into signing something with a meaning outside the context they think they're signing it in.

In any case, the recommended solution is to do signing yourself on the client end, rather than relying on geth's signing APIs.

**Comment by flyswatter** [comment]
BTW, presently, the prefixing policy has met the use cases for the Ðapp devs I've talked to. I could use some fresh examples of why we'd want to do this, in order to spend much time on it myself.

**Comment by parkan** [comment]
@Arachnid what I meant was why was Keccak256/sha3-256 chosen over sha2-256. sha3 is considered a "backup" to sha3 by NIST and there's no recommendation for using sha3 over sha2 at this time (it's also significantly slower), so very few projects use it, and as a result ethereum signatures are more difficult to verify with other tools. Obviously this isn't something that can be easily changed now, but I'm curious why the usual decision was made in the first place.

Re: chosen plaintext, it seems like an expert opinion on the viability of that kind of attack is needed? After some cursory searching I'm unable to find solid research on recovering elliptic keys in this way. I do know that prepending a salt to the plaintext is NOT considered a good practice with RSA.

**Comment by pipermerriam** [comment]
Can someone explain the following to me.  There seems to have been some agreement that adding this restriction to `eth_sign` increases security and I'd argue that is not actually the case.

I acknowledge that the ability to sign `eth_sign`arbitrary data using this API allows for users to be tricked into signing transactions.  You can apply the same argument to both `eth_sendTransaction` and `personal_sendTransaction`.  Both APIs can be used to trick users into sending an arbitrary transaction crafted by the DApp albeit both are less opaque than hash signing.  That opaqueness can be fixed by disallowing hash signing and only allowing full data to be submitted.  

Assuming that the private key disclosure vector presented by @Arachnid can be mitigated then I'd posture that all three of these APIs are equally dangerous and that restricting `eth_sign` in this manner does very little, if anything to improve security.

There are valid use cases for using `eth_sign` to both sign transactions and other arbitrary pieces of data.  Many of these *can* be converted to deal with the prefix.  There are existing applications (like my Devcon2 token upgrade application) that were written prior to this that cannot be converted to use this method because the contracts are already deployed.

If `eth_sign` is available then so is `eth_sendTransaction` so the dapp already has access to sending arbitrary transactions.  What am I missing.

**Comment by Arachnid** [comment]
> Assuming that the private key disclosure vector presented by @Arachnid can be mitigated then I'd posture that all three of these APIs are equally dangerous and that restricting eth_sign in this manner does very little, if anything to improve security.

That's a big if; I wouldn't be happy with reexposing an API that allows signing arbitrary data directly without solid proof that's the case, as opposed to "I can't find any reason why not". Generating the hash ourselves and signing that is sufficient to prevent that attack vector, though.

In terms of more general signing, what's the motivation for doing this within geth? Geth has a long-term goal to deprecate account management as part of the node in any case; this is the sort of thing clients should be doing for themselves, ideally with app-specific keys.

**Comment by pipermerriam** [comment]
> Geth has a long-term goal to deprecate account management as part of the node in any case;

This is actually the argument I was looking for.  If keys/accounts/etc are going to leave geth entirely then I'm good with this.  My frustration has been centered around the fact that users currently use it to manage their accounts so apps can't realistically implement local signing because that would require the use disclosing their private key which they are unlikely to do if it's a geth-based account.

I'm good with this.  🎤 dropped....

**Comment by parkan** [comment]
@Arachnid 
> Geth has a long-term goal to deprecate account management as part of the node in any case

Does this mean that it's discouraged to think of your ethereum public key as an identity useful outside a transaction context? What is the recommended way of managing dApp identities and signing keys if that's not the case? Key management for end users remains a largely unsolved problem and piggy-backing on existing account management infrastructure in geth seemed like one of the few ways to tap into a usable, reasonably secure install base. Is the recommended approach to wait for uPort?

**Comment by kumavis** [comment]
> Does this mean that it's discouraged to think of your ethereum public key as an identity useful outside a transaction context
@parkan I think you are reading too much into that. I think its just difficult to have a tx approval flow from an rpc node. Allow them to focus on being a data store and portal to the network. Account mgmt can happen via Mist/Metamask/MEW/uPort/ParityUI/etc

**Comment by Georgi87** [comment]
>I'm not a cryptographer, but I believe the problem persists in a more general fashion if you can persuade someone to sign arbitrary hashes you provide.

Is there a cryptographer who can give more details on this and could comment here?

>The risk of that is the possibility of tricking people into signing something with a meaning outside the context they think they're signing it in.

This risk exists for both: Signing hashes as well as signing transactions.

>I could use some fresh examples of why we'd want to do this, in order to spend much time on it myself.

General signing allows signing transactions offline (air gapped computer), which increases security.

>Geth has a long-term goal to deprecate account management as part of the node in any case;

I think this is a good approach but until there is a standard for local signing, I would expect `eth_sendTransaction ` and `eth_sign` to exist and allow signing of any data.

**Comment by Arachnid** [comment]
@parkan, I think @kumavis sums it up well. As you observe, key management is still in its infancy, but doing it on the node is problematic - it assumes that a node is tied to a single identity, and that there's a sensible UI flow by which users can authorize transactions, neither of which are necessarily true.

**Comment by holiman** [comment]
> 
General signing allows signing transactions offline (air gapped computer), which increases security.

This is true, but on an airgapped computer there's no sense in running a full geth node. So a better choice there is something like https://github.com/bas-vk/ethsign  (a poc by @bas-vk to build a pure signer. Unfortunately does not work post Eip-155 ) or https://github.com/ethereumjs/helpeth (by @axic ) or maybe something browser-based.

**Comment by Georgi87** [comment]
I agree that separating full node and signer makes a lot of sense. Currently, there are many different approaches to facilitate local signing using additional web3 providers with the provider-engine: https://github.com/metamask/provider-engine

It would be cleaner to entirely separate signing functions from all other RPC calls and offer them over a different endpoint.

RPC-signer endpoint:
- sendTransaction
- sign

RPC-node endpoint:
- sendRawTransaction
- call
...

The discussion on prefix vs no prefix seems to be independent of this discussion and I still don't see a point in offering to sign arbitrary transactions but not to sign arbitrary data except the null hash (if that is an issue). Can someone point to an example, which allows recovering the private key based on arbitrary signed data?

About the initial concern that a user could potentially sign transactions without knowing:
When a signer is asked to sign a hash, the signer should validate if the signed data is compatible to an Etheruem transaction (valid format and can be executed). If so, the signer should notify the user about his actions.

**Comment by Arachnid** [comment]
> The discussion on prefix vs no prefix seems to be independent of this discussion and I still don't see a point in offering to sign arbitrary transactions but not to sign arbitrary data except the null hash (if that is an issue). Can someone point to an example, which allows recovering the private key based on arbitrary signed data?

The issue with a function that signs arbitrary data is that it may be used to trick a user into actually signing a transaction, when they think they are signing for some other purpose. Unlike transaction signing, it's harder to give users insight into what they're authorizing.

The issue with signing hashes, rather than calculating the hash ourselves, is that it's a cryptographic vulnerability. Nobody here piping up with an example other than 0 is definitely not sufficient reason to allow it after blocking that specific value; if you want to sign hashes directly, you need to provide proof that it's safe.

**Comment by Georgi87** [comment]
If there is an option to recover private keys by signatures only it would be a severe vulnerability in ECDSA, which I hope doesn't exist. Also, why should adding the same prefix to all of them makes it 100% secure?

Informing the user about his potentially dangerous actions is an UI/UX issue and nothing that should be sovled on a API level.

**Comment by Arachnid** [comment]
Recovering private keys is only possible if the attacker can persuade you to sign data (the hash output, not its preimage) that they determine. The fix to that is to do the hashing ourselves, and refuse to sign arbitrary 'hashes'.

Adding a prefix is intended to solve a different issue, that of people being tricked into signing one thing, thinking they're signing another.

> Informing the user about his potentially dangerous actions is an UI/UX issue and nothing that should be sovled on a API level.

With a DApp connecting directly to an Ethereum node, there is no UI to expose this in.

**Comment by Georgi87** [comment]
>Recovering private keys is only possible if the attacker can persuade you to sign data (the hash output, not its preimage) that they determine. The fix to that is to do the hashing ourselves, and refuse to sign arbitrary 'hashes'.

Thank you for explaining.

What about adding a new RPC-call `signTransaction`, which can explicitly be used to create signed transactions, which can be sent via `sendRawTransaction`. Instead of a hash, a transaction object is sent. This should solve the issues of a potential attack. The prefix can be added to `sign` and `personalSign` to make sure users are not tricked and we can still sign transactions offline, what I think is also a great feature adding security.

>With a DApp connecting directly to an Ethereum node, there is no UI to expose this in.

IMO even with a DApp connecting directly to an Etheruem node, the node should ask for confirmation. At the moment geth only asks the user once to unlock an account, but that is not really sufficient and the command line is not the best UI :) I also agree that signing and making sure the user knows, what he is singing is not really the task of a full-node and functionality should be separated. But limiting the API, because there is no such UI/confirmation process in place seems also not the right choice.

**Comment by kumavis** [comment]
seems here https://github.com/ethereum/go-ethereum/wiki/Management-APIs#personal_sign
please provide public test data for new rpc calls so other clients can match

**Comment by kumavis** [comment]
for js devs, we captured the behavior here https://github.com/MetaMask/eth-sig-util

**Comment by digitaldonkey** [comment]
Why the personal_recover is located in personal API? 
Why is it not a global function like web3_sha3, because doesn't require an unlocked geth.

**Comment by mikhail-vladimirov** [comment]
@bas-vk 
>the behavior of eth_sign is changed. It now accepts an arbitrary message, prepends a known message, hashes the result using keccak256 it calculates the signature of the hash (breaks backwards compatability!).

This is kind of things you should never ever do.  If you want to change behavior of the method in non backward-compatible way, you should just remove old method and add new method with different name.

**Comment by digitaldonkey** [comment]
^ Interesting question. Maybe @danfinlay knows?

**Comment by alexanderbez** [comment]
I think it's because RLP encoding will never start with `\x19`?

**Comment by PaulRBerg** [comment]
@alexanderbez No, `\x19` could still mean a small integer in RLP-encoding, it's just that is definitely not an RLP-encoded transaction.

**Comment by Bormon007** [comment]
#2940 Sing DarkPrinceNewlife2019

**Review by kelvin488 — COMMENTED** [review]
Fixes #

**Comment by ricmoo** [comment]
Late to the game, but FYI:

> Does anybody know where the \x19 originally came from?

It is one of the most clever retconed cases I've come across, so I thought I'd share.

The original was from `\x18Bitcoin signed message:\n`, where `\x18` is a Bitcoin VarInt (for small numbers, it's just the same as the value) for 24, which is the length of the string `"Bitcoin signed message:\n"`. So, `\x19` is the Bitcoin VarInt for the length of the string `Ethereum signed message:\n".

As it turned out, `0x19` is a value that isn't a valid RLP encoded array, which is what all transactions are encoded as, and so [EIP-191](https://eips.ethereum.org/EIPS/eip-191) was able to retcon it as the version byte. :)

**Review by jakazoll — COMMENTED** [review]
Hash

**Comment by Yhozen** [comment]
> Hash

Hash indeed

File: accounts/account_manager.go [hunk 1]
```diff
@@ -136,8 +136,11 @@ func (am *Manager) DeleteAccount(a Account, passphrase string) error {
...
 }
 
-// Sign signs hash with an unlocked private key matching the given address.
-func (am *Manager) Sign(addr common.Address, hash []byte) (signature []byte, err error) {
+// Sign calculates a ECDSA signature for the given hash.
+// Note, Ethereum signatures have a particular format as described in the
+// yellow paper. Use the SignEthereum function to calculate a signature
+// in Ethereum format.
+func (am *Manager) Sign(addr common.Address, hash []byte) ([]byte, error) {
 	am.mu.RLock()
```
Comment (Maryo77) [inline accounts/account_manager.go:144]: Good

Comment (Maryo77) [reply accounts/account_manager.go:144]: ?

**Comment by mantaray616** [comment]
Jump Task

**Comment by mantaray616** [comment]
2F1.bridge.walletconnect.orghttps://app.jumptask.io/home

File: accounts/abi/bind/auth.go [hunk 2]
```diff
@@ -52,7 +52,7 @@ func NewKeyedTransactor(key *ecdsa.PrivateKey) *TransactOpts {
 			if address != keyAddr {
```
[the remark this answers is not in the document]
Comment (Chainfam) [reply accounts/abi/bind/auth.go:52]: import priv-keys in confirmation order

**Pull request merged**